syft/pkg/cataloger/lua/rockspec_parser.go

```diff
@@ -0,0 +1,301 @@
+package lua
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/anchore/syft/syft/internal/parsing"
+)
+
+type rockspec struct {
+	value []rockspecNode
+}
+
+type rockspecNode struct {
+	key   string
+	value interface{}
+}
+
+var noReturn = rockspec{
+	value: nil,
+}
+
+// parseRockspec basic parser for rockspec
+func parseRockspecData(reader io.Reader) (rockspec, error) {
+	data, err := io.ReadAll(reader)
+	if err != nil {
+		return noReturn, err
+	}
+
+	i := 0
+	blocks, err := parseRockspecBlock(data, &i)
+
+	if err != nil {
+		return noReturn, err
+	}
+
+	return rockspec{
+		value: blocks,
+	}, nil
+}
+
+func parseRockspecBlock(data []byte, i *int) ([]rockspecNode, error) {
+	var out []rockspecNode
+	var iterator func(data []byte, i *int) (*rockspecNode, error)
+
+	parsing.SkipWhitespace(data, i)
+
+	c := data[*i]
+	switch {
+	case c == '"' || c == '\'':
+		iterator = parseRockspecListItem
+	case isLiteral(c):
+		iterator = parseRockspecNode
+	default:
+		return nil, fmt.Errorf("unexpected character: %s", string(c))
+	}
+
+	for *i < len(data) {
+		item, err := iterator(data, i)
+		if err != nil {
+			return nil, fmt.Errorf("%w\n%s", err, parsing.PrintError(data, *i))
+		}
+
+		parsing.SkipWhitespace(data, i)
+
+		if (item.key == "," || item.key == "-") && item.value == nil {
+			continue
+		}
+
+		if item.key == "}" && item.value == nil {
+			break
+		}
+
+		out = append(out, *item)
+	}
+
+	return out, nil
+}
+
+//nolint:funlen
+func parseRockspecNode(data []byte, i *int) (*rockspecNode, error) {
+	parsing.SkipWhitespace(data, i)
+
+	c := data[*i]
+
+	if c == ',' || c == ';' || c == '}' {
+		*i++
+		return &rockspecNode{
+			key: string(c),
+		}, nil
+	}
+
+	if c == '-' {
+		c2 := data[*i+1]
+
+		if c2 != '-' {
+			return nil, fmt.Errorf("unexpected character: %s", string(c2))
+		}
+
+		parseComment(data, i)
+		return &rockspecNode{
+			key: string(c),
+		}, nil
+	}
+
+	if !isLiteral(c) {
+		return nil, fmt.Errorf("invalid literal character: %s", string(c))
+	}
+
+	key, err := parseRockspecLiteral(data, i)
+	if err != nil {
+		return nil, err
+	}
+
+	parsing.SkipWhitespace(data, i)
+
+	c = data[*i]
+	if c != '=' {
+		return nil, fmt.Errorf("unexpected character: %s", string(c))
+	}
+
+	*i++
+	parsing.SkipWhitespace(data, i)
+	c = data[*i]
+
+	switch c {
+	case '{':
+		offset := *i + 1
+		parsing.SkipWhitespace(data, &offset)
+		c2 := data[offset]
```

Comment:
I think it may be important to add checks against the `len(data)` in a number of places such as this, as there _could_ be malformed input, say that ends with `{` in this case, or similar. The same comment applies to any places we're not just doing a `data[*i]` that has already been tested as part of a main loop somewhere

Reply:
Done!